### PR TITLE
feature: Issue #17 - Registration Number Policies

### PR DESCRIPTION
**Pull request by PooriaT**
## Description

Issue Link:  [Mock policies for registration number #17](https://github.com/OpenBCca/airbnb-regulation/issues/17)

In this issue, a verification for the registration number is provided. First, the registration number has to be validated. Then, it should be checked whether it is available in Vancouver Open Data. 

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

- [x] Unit Test
- [ ] Integration Test
- [ ] E2E Test

## Checklist:

Before you submit your pull request, please make sure you have completed the following:

- [ ] I have read the CONTRIBUTING document.
- [x] I have checked that my code adheres to the code style of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: analytics/policy/services/registration_number_policies.py
Comment (nam-m) [inline analytics/policy/services/registration_number_policies.py:1]: The policies should not be in services folder

File: analytics/policy/services/registration_number_policies.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import json
+import urllib3
+
+class RegistrationNumberPolicies:
+    def __init__(self, registration_number):
+        self.registration_number = registration_number
+        self.session = urllib3.PoolManager()
+
+    def valid_pattern(self):
+        valid_registration_pattern = re.compile(r'^[0-9]{2}-[0-9]{6}$')
```
Comment (nam-m) [inline analytics/policy/services/registration_number_policies.py:11]: The regex should check first two digits against year end digits.
Please check issue description

File: analytics/policy/services/registration_number_policies.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import re
+import json
+import urllib3
+
+class RegistrationNumberPolicies:
+    def __init__(self, registration_number):
+        self.registration_number = registration_number
```
Comment (nam-m) [inline analytics/policy/services/registration_number_policies.py:7]: I think the registration number should be retrieved from listing model.
Please check analytics/models/listing.py

File: analytics/policy/services/registration_number_policies.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import re
+import json
+import urllib3
```
Comment (nam-m) [inline analytics/policy/services/registration_number_policies.py:3]: No need to make request with urllib3 in the policies, since this is handled in API client

File: analytics/policy/services/registration_number_policies.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        facet_name = "licencenumber"
+        limit_value = 20
+        OPEN_DATA_VANCOUVER_URL = (
+            "https://opendata.vancouver.ca/"
+            "api/explore/v2.1/catalog/datasets/"
+            "business-licences/records?"
+            f"limit={limit_value}&"
+            f"refine={facet_name}%3A{self.registration_number}"
+        )
+        response = self.session.request(
```
Comment (nam-m) [inline analytics/policy/services/registration_number_policies.py:28]: There's no need to access the API as this is handled in the API client, please check PR #38 (this is still under review, so let me know if you have any concerns). You can call the client here
The return values from the client will be in the form of ["status"], ["status1", "status2"], ... Please check the API client unit test in the same PR for examples

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import pytest
+from unittest.mock import MagicMock
+from policy.services.registration_number_policies import RegistrationNumberPolicies
+
+@pytest.fixture
+def mock_http_response_200():
```
Comment (nam-m) [inline analytics/policy/tests/unit/test_registration_number_policies.py:6]: No need to mock http response

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    response_mock = MagicMock()
+    response_mock.status = 200
+    response_mock.data = '{"total_count": 1}'
+    return response_mock
+
+@pytest.fixture
+def mock_http_no_response():
```
Comment (nam-m) [inline analytics/policy/tests/unit/test_registration_number_policies.py:13]: No need to mock http response

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        "q4-160188",
+        "4-160188",  
+    ]
+    for registration_number in registration_numbers:
+        registration_number_policies = RegistrationNumberPolicies(
+            registration_number
+        )
+        assert registration_number_policies.valid_pattern() is False
+
+def test_valid_registration_number(mock_http_response_200):
```
Comment (nam-m) [inline analytics/policy/tests/unit/test_registration_number_policies.py:42]: No need to test mock http response

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    registration_number = "24-160188"
+    registration_number_policies = RegistrationNumberPolicies(
+        registration_number
+    )
+    registration_number_policies.session.request = MagicMock(
+        return_value=mock_http_response_200
+    )
+    assert registration_number_policies.valid_registration_number() is True
+
+def test_invalid_registration_number(mock_http_no_response):
```
Comment (nam-m) [inline analytics/policy/tests/unit/test_registration_number_policies.py:52]: No need to test mock http response

**Review by nam-m — CHANGES_REQUESTED** [review]
I think the policy structure is good, and it's very readable.
You don't need to call the API in here though, because it is handled in #38

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+def test_invalid_pattern():
+    registration_numbers = [
+        "24160188",
+        "24-160a88",
+        "24-1601883",
+        "24-1601a8",
+        "124-160188",
+        "q4-160188",
+        "4-160188",
```
Comment (umsu2) [inline analytics/policy/tests/unit/test_registration_number_policies.py:34]: other patterns include None, "", 123 etc.

File: analytics/policy/services/registration_number_policies.py
Comment (PooriaT) [reply analytics/policy/services/registration_number_policies.py:1]: Which directory do you recomment?

File: analytics/policy/services/registration_number_policies.py (hunk 2 above)
Comment (PooriaT) [reply analytics/policy/services/registration_number_policies.py:7]: Don't you want to use registration_number_poiclies as a helper to check the validity of the license number?
In this case Listing will be imported in another part of the code.

File: analytics/policy/services/registration_number_policies.py (hunk 3 above)
Comment (PooriaT) [reply analytics/policy/services/registration_number_policies.py:3]: So, it means that I need to remove the second method which is valid_registration_number.

File: analytics/policy/services/registration_number_policies.py
Comment (nam-m) [reply analytics/policy/services/registration_number_policies.py:1]: policy/policies?

File: analytics/policy/services/registration_number_policies.py (hunk 2 above)
Comment (nam-m) [reply analytics/policy/services/registration_number_policies.py:7]: I mean the attribute `registration_number` should be taken from Listing object
But this maybe fine for now for simplicity

File: analytics/policy/services/registration_number_policies.py (hunk 3 above)
Comment (nam-m) [reply analytics/policy/services/registration_number_policies.py:3]: no, it means you can use the BusinessLicenceClient object to retrieve whatever you want to check with the given `registration_number`

File: analytics/policy/services/registration_number_policies.py (hunk 4 above)
Comment (nam-m) [reply analytics/policy/services/registration_number_policies.py:28]: with respect to Yang's comment, you can make it like:

ValidRegistrationNumberPolicy(Policy)
UniqueRegistrationNumberPolicy(Policy)
ExistedRegistrationNumberPolicy(Policy)

where Policy is the interface with any shared method/attribute
These Policy objects mimic the strategy design pattern

File: analytics/models/listing.py [hunk 10]
```diff
@@ -22,3 +22,4 @@ def __init__(self, id: str, *, address: Address, licence_number: str = None):
         self.id = id
         self.address = address
         self.licence_number = licence_number
+
```
Comment (nam-m) [inline analytics/models/listing.py:25]: please include new line

File: analytics/policy/policies/registration_number_policies.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            if int(year) > END_YEAR or int(year) < START_YEAR:
+                return False
+            else:
+                return True
+        else:
+            return False
+
+    def unique_registration_number_policy(self):
+        business_licences = BusinessLicenceClient()
+        licence_status, total_count = business_licences.get_licence_status(
```
Comment (nam-m) [inline analytics/policy/policies/registration_number_policies.py:28]: this method only returns `licence_status` as a list of str, so you don't need `total_count` and can check list length instead

File: analytics/policy/policies/registration_number_policies.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+            if licence_status[0] == "Issued":
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    def existed_registration_number_policy(self):
+        business_licences = BusinessLicenceClient()
+        licences_status, total_count = business_licences.get_licence_status(
```
Comment (nam-m) [inline analytics/policy/policies/registration_number_policies.py:41]: same as previous comment

File: analytics/policy/services/business_licence_client.py [hunk 13]
```diff
@@ -67,7 +67,7 @@ def get_licence_status(self, licence_number: str) -> list[str]:
         )
         response_data = self._make_request(fields)
 
-        return self._process_licence_status_results(response_data)
+        return self._process_licence_status_results(response_data), response_data["total_count"]
```
Comment (nam-m) [inline analytics/policy/services/business_licence_client.py:70]: this method should only return statuses, so if you want to check number of statuses, please use length of returned list

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 14]
```diff
@@ -31,30 +16,37 @@ def test_invalid_pattern():
...
+    registration_number_policies = RegistrationNumberPolicies(registration_number) 
+    assert registration_number_policies.unique_registration_number_policy() is True # (["Issued"], 1)
+
+    registration_number = "20-160574"
+    registration_number_policies = RegistrationNumberPolicies(registration_number) 
+    assert registration_number_policies.unique_registration_number_policy() is False # (["Pending", "Inactive"], 1)
+
+    # registration_number = ""
+    # registration_number_policies = RegistrationNumberPolicies(registration_number)
+    # assert registration_number_policies.unique_registration_number_policy() is False # (["Issued", "Issued"], 2)
```
Comment (nam-m) [inline analytics/policy/tests/unit/test_registration_number_policies.py:41]: total_count should not be in the comments

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 15]
```diff
@@ -31,30 +16,37 @@ def test_invalid_pattern():
...
     ]
     for registration_number in registration_numbers:
         registration_number_policies = RegistrationNumberPolicies(
             registration_number
         )
-        assert registration_number_policies.valid_pattern() is False
+        assert registration_number_policies.valid_registration_number_policy() is False
 
-def test_valid_registration_number(mock_http_response_200):
+def test_unique_registration_number_policy():
```
Comment (nam-m) [inline analytics/policy/tests/unit/test_registration_number_policies.py:30]: please include test cases for registration_number = "", None

File: analytics/policy/tests/unit/test_registration_number_policies.py [hunk 16]
```diff
@@ -31,30 +16,37 @@ def test_invalid_pattern():
...
+
+def test_existed_registration_number_policy():
+
+    registration_number = "24-160188"
+    registration_number_policies = RegistrationNumberPolicies(registration_number)
+    assert registration_number_policies.existed_registration_number_policy() is True
+
+    # assert registration_number_policies.existed_registration_number_policy() is False
+
+    # assert registration_number_policies.existed_registration_number_policy() is False
```
Comment (nam-m) [inline analytics/policy/tests/unit/test_registration_number_policies.py:51]: please remove not needed comments

**Review by nam-m — CHANGES_REQUESTED** [review]
Thanks for your update

2 changes I recommend are:
1. Remove `total_count` check in the policies
2. In the tests, you can set registration numbers one time outside(maybe parametrize too),  and let all tests go through these

File: analytics/policy/tests/unit/test_registration_number_policies.py (hunk 15 above)
Comment (PooriaT) [reply analytics/policy/tests/unit/test_registration_number_policies.py:30]: These two are tested in `test_invalid_pattern`

Comment (nam-m) [reply analytics/policy/tests/unit/test_registration_number_policies.py:30]: yeah, but the unique and existed policies should be robust enough to handle these

File: analytics/policy/policies/registration_number_policies.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+            return False
+
+    def unique_registration_number_policy(self):
+        business_licences = BusinessLicenceClient()
+        licence_status = business_licences.get_licence_status(
+            self.registration_number
+        )
+        total_count = len(licence_status)
+        if total_count == 1:
+            if licence_status[0] == "Issued":
```
Comment (nam-m) [inline analytics/policy/policies/registration_number_policies.py:33]: sorry i forgot, the licence status result should not be checked here.
it shouldn't matter whether it's "Issued", "Cancelled",... as only uniqueness is checked

File: analytics/policy/policies/registration_number_policies.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+            return False
+
+    def existed_registration_number_policy(self):
+        business_licences = BusinessLicenceClient()
+        licences_status = business_licences.get_licence_status(
+            self.registration_number
+        )
+        total_count = len(licences_status)
+        if total_count > 0:
+            if all(status == "Issued" for status in licences_status):
```
Comment (nam-m) [inline analytics/policy/policies/registration_number_policies.py:47]: same as previous comment

**Review by nam-m — CHANGES_REQUESTED** [review]

**Review by nam-m — APPROVED** [review]

**Review by nam-m — APPROVED** [review]
